src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,384 @@
+    };
+    setResult(result);
+}
+
+void SearchController::resultsAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    int limit = params()["limit"].toInt();
+    int offset = params()["offset"].toInt();
+
+    auto searchHandlers = sessionManager()->session()->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+    const QList<SearchResult> searchResults = searchHandler->results();
+    const int size = searchResults.size();
+
+    if ((offset > size) || ((offset == size) && (offset != 0)) || (limit < 0))
+        throw APIError(APIErrorType::Conflict, tr("Offset or limit is out of range"));
+
+    // normalize values
+    if (offset < 0)
+        offset = size + offset;
+    if (offset < 0)  // check again
+        throw APIError(APIErrorType::Conflict);
+    if (limit == 0)
+        limit = -1;
+
+    if ((limit > 0) || (offset > 0))
+        setResult(getResults(searchResults.mid(offset, limit), searchHandler->isActive(), size));
+    else
+        setResult(getResults(searchResults, searchHandler->isActive(), size));
+}
+
+void SearchController::deleteAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    ISession *session = sessionManager()->session();
+
+    auto searchHandlers = session->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+    searchHandler->cancelSearch();
+    searchHandlers.remove(id);
+    session->setData(QLatin1String("searchHandlers"), QVariant::fromValue(searchHandlers));
+
+    auto activeSearches = session->getData(QLatin1String("activeSearches")).value<QSet<QString>>();
+    activeSearches.remove(id);
+    session->setData(QLatin1String("activeSearches"), QVariant::fromValue(activeSearches));
+}
+
+void SearchController::categoriesAction()
+{
+    QStringList categories;
+    const QString name = params()["pluginName"].trimmed();
+
+    categories << SearchPluginManager::categoryFullName("all");
+    for (const QString &category : SearchPluginManager::instance()->getPluginCategories(name))
+        categories << SearchPluginManager::categoryFullName(category);
+
+    const QJsonArray result = QJsonArray::fromStringList(categories);
+    setResult(result);
+}
+
+void SearchController::pluginsAction()
+{
+    const QStringList allPlugins = SearchPluginManager::instance()->allPlugins();
+    setResult(getPluginsInfo(allPlugins));
+}
+
+void SearchController::installPluginAction()
+{
+    checkParams({"sources"});
+
+    const QStringList sources = params()["sources"].split('|');
+    for (const QString &source : sources)
+        SearchPluginManager::instance()->installPlugin(source);
```

Comment:
Can I reach DoS state invoking this frequently enough?

Reply:
Is this function more vulnerable than any other api endpoint?

Reply:
Those two functions (update and install) consume resources and free them only after downloading (successfully or not) a file. If downloading speed is low, the resources can be acquires as many times as the API calls were made.

Reply:
In that case I suppose it's theoretically possible, though these APIs are CSRF-protected POSTs. Only the user can invoke them.

Reply:
>these APIs are CSRF-protected POSTs. Only the user can invoke them.

It's wrong assumption.
There are no POST, GET, etc. on this app level. It's not Web API. It's just "serial" API, that gets serialized data, process it and returns serialized result.
If you need some WebAPI related limitations you need to implement them one level up.

Reply:
I am referring to the Web API being CSRF-protected.

Reply:
I just thought you were focusing on the request method...